### PR TITLE
Report for animal abilities in shelter

### PR DESCRIPTION
**Pull request by Kamen06**

### CODE REVIEW

File: src/animalShelter/AnimalShelter.ts [hunk 1]
```diff
@@ -68,8 +69,33 @@ export default class AnimalShelter {
...
+    for(const animal of this.animals){
+      if(this.hasMethod(animal, "jump")){
+        animalsWhoJump++
+      }
+      if(this.hasMethod(animal, "run")){
+        animalsWhoRun++;
+      } 
+      if(animal._canClimbTrees){
+        anaimalsWhoClimbTrees++;
+      }
```
Comment (teodora1203) [inline src/animalShelter/AnimalShelter.ts:92]: either use the hasMethod for every activity or add properties like the canClimbTrees, using both ways is inconsistent. also shouldn't the animals who can climb trees actually have a method `climb`?

File: src/hierarchy/Animal.ts [hunk 2]
```diff
@@ -15,6 +15,7 @@ export default abstract class Animal implements ICanWalk {
   protected home?: string;
   protected timesGivenBirth: number;
   protected _canHaveFamily: boolean = false;
+  public _canClimbTrees = false;
```
Comment (teodora1203) [inline src/hierarchy/Animal.ts:18]: the underscore `_` is usually used only for private fields

File: src/animalShelter/AnimalShelter.ts [hunk 3]
```diff
@@ -68,8 +69,33 @@ export default class AnimalShelter {
       console.log(
         `Cannot change reptile limit to ${newReptileLimit} there ${reptilesTotal == 1 ? "is" : "are"} already ${reptilesTotal} ${reptilesTotal == 1 ? "reptile" : "reptile"}`
       );
+    };
+  };
+
+  private hasMethod(animal:Animal, methodName:string): boolean{
+    return typeof (animal as any)[methodName] === "function";
```
Comment (teodora1203) [inline src/animalShelter/AnimalShelter.ts:76]: you shouldn't cast to any

File: src/animalShelter/AnimalShelter.ts (hunk 1 above)
Comment (rkoychev) [reply src/animalShelter/AnimalShelter.ts:92]: We only have the knowledge if it climbs or not, that is why this is a boolean property and the others are mthods with activities

File: src/animals/Crocodile.ts [hunk 4]
```diff
@@ -9,13 +9,15 @@ export default class Crocodile extends Reptile {
...
 
+
   constructor(name: string, age: number, isMale: boolean, length: number) {
     super(name, age, isMale);
     this._canHaveFamily = true;
     if (length <= 0) {
       throw new Error("Length must be greater than zero");
     }
     this.length = length;
+    this._canClimbTrees =true;
```
Comment (rkoychev) [inline src/animals/Crocodile.ts:20]: not formated

File: src/animals/Lion.ts [hunk 5]
```diff
@@ -11,6 +11,7 @@ export default class Lion extends Mammal implements ICanRun, ICanTalk, ICanJump
   constructor(name: string, age: number, isMale: boolean) {
     super(name, age, isMale);
     this._canHaveFamily = true;
+    this._canClimbTrees =true;
```
Comment (rkoychev) [inline src/animals/Lion.ts:14]: not formated

File: src/animals/Snake.ts [hunk 6]
```diff
@@ -27,6 +27,7 @@ export default class Snake extends Reptile {
     this.type = type;
     this.color = color;
     this.length = length;
+    this._canClimbTrees =true;
```
Comment (rkoychev) [inline src/animals/Snake.ts:30]: not formated

File: src/animals/Squirrel.ts [hunk 7]
```diff
@@ -27,6 +27,7 @@ export default class Squirrel extends Mammal implements ICanJump {
     this.treeType = treeType;
     this.treeAge = treeAge;
     this.holeSize = holeSize;
+    this._canClimbTrees =true;
```
Comment (rkoychev) [inline src/animals/Squirrel.ts:30]: not formated

File: src/animals/Tortoise.ts [hunk 8]
```diff
@@ -11,6 +11,7 @@ export default class Tortoise extends Reptile {
     super(name, age, isMale);
 
     this.home = "Animal Kingdom";
+    this._canClimbTrees =true;
```
Comment (rkoychev) [inline src/animals/Tortoise.ts:14]: not formated

File: src/animalShelter/AnimalShelter.ts (hunk 3 above)
Comment (Kamen06) [reply src/animalShelter/AnimalShelter.ts:76]: Usually yes, but in this case, in my opinion, we are sure it is an animal since it has been passed as one

Comment (rkoychev) [reply src/animalShelter/AnimalShelter.ts:76]: @Kamen06  still not a reason to cast to "any"

Comment (Kamen06) [reply src/animalShelter/AnimalShelter.ts:76]: i cannot think of another way of checking if an object has a specific function

Comment (teodora1203) [reply src/animalShelter/AnimalShelter.ts:76]: I still think we shouldn't be using this `hasMethod`, it's a weird way of reporting the animal's abilities. @rkoychev isn't it better to add properties `canJump` and `canRun` the same way we have for climbing the trees?

File: src/animalShelter/AnimalShelter.ts [hunk 9]
```diff
@@ -72,19 +74,15 @@ export default class AnimalShelter {
...
-    return typeof (animal as any)[methodName] === "function";
-  };
-
   public report() {
     let animalsWhoJump = 0;
     let animalsWhoRun = 0;
     let anaimalsWhoClimbTrees = 0
     for (const animal of this.animals) {
-      if (this.hasMethod(animal, "jump")) {
+      if ((animal as unknown as ICanJump).jump) {
```
Comment (teodora1203) [inline src/animalShelter/AnimalShelter.ts:82]: casting to unknown and then to ICanJump is also not a great solution. If you saw in the group chat Rosen posted a solution: in the `hasMethod` method you can `return animal.constructor.property.hasOwnProperty(methodName)` I believe that works but test it yourself too

**Review by rkoychev — APPROVED** [review]

**Pull request merged**